### ISSUE TITLE
Log retention not configurable in helm chart

### ISSUE THREAD

**Issue by deimosfr** [grafana/loki#9207]
**Is your feature request related to a problem? Please describe.**
Hi,

I recently updated Loki chart and saw a lot of merges of multiple charts in one, leading to configuration changes in the Values.yaml.

Since then, I observed that the config part is less flexible than before. The missing part in the new chart is the ability to configure the retention period, so we can't control it directly with the chart.

**Describe the solution you'd like**
If you look [at the chart](https://github.com/grafana/loki/blob/main/production/helm/loki/values.yaml#L148-L150), you'll find:
```yaml
    table_manager:
      retention_deletes_enabled: false
      retention_period: 0
```
Where it completely disables it. However, I'm using the single binary version and I need to tune this part.

It would be great to have configurable parameters like other fields. For example:
```
    {{- with .Values.loki.table_manager }}
    table_manager:
      {{- tpl (. | toYaml) $ | nindent 4 }}
    {{- end }}
```

Are there any reasons behind it why it has not been done? If it's just something that has been forgotten, I can make a PR if you're ok.

Thanks in advance!

**Comment by nchal** [comment]
I came to the same solution than u @deimosfr  
And surprisingly, there is not that much urge on that topic... Does ppl not configure the retention ? 😔

**Comment by khanh96le** [comment]
For the chart version `4.9.2`, this works for me:
* set `loki.compactor.retention_enabled`  to true
* and set the `loki.limits_config.retention_period` value

**Comment by deimosfr** [comment]
Thanks!

**Comment by spuyet** [comment]
I also had to configure:
```
delete_request_store: 'filesystem'
```
 to make it work.

**Comment by yahav2305** [comment]
> I also had to configure:
> 
> ```
> delete_request_store: 'filesystem'
> ```
> 
> to make it work.

Where did you specify this? An example with a helm file would be appreciated

**Comment by younsl** [comment]
Here is my solution tested on EKS v1.28 with [loki-distributed](https://github.com/grafana/helm-charts/tree/main/charts/loki-distributed) 0.79.2 (loki v2.9.8) helm chart.

---

## Overview

This guide explains how to set up log retention using the compactor in the Loki-distributed chart.

## Environment

- **Kubernetes Cluster**: EKS v1.28 (amd64)
- **Helm chart**: [loki-distributed](https://github.com/grafana/helm-charts/tree/main/charts/loki-distributed) v0.79.2 (appVersion: loki v2.9.8)
- **Loki storage type**: persistentVolume backed by gp3 type EBS storageClass

## Background Knowledge

### Log retention in Loki

Retention in Grafana Loki is managed either through the Table Manager or the Compactor.

> [!WARNING]
> **Comparison of compactor and table-manager**
>
> In `loki-distributed` chart v0.79.2 (loki v2.9.8), the `table-manager` is disabled by default, and its configuration is entirely omitted from `Values.loki.config`. The `table-manager` is no longer used (deprecated) and is not recommended for new deployments. Its functionalities have now been replaced by the `compactor`, which is designed to handle log retention and table management, among other features.

By default, if the `table_manager.retention_deletes_enabled` or `compactor.retention_enabled` flags are not set, logs sent to Loki will be retained indefinitely.

For more details, refer to the [Retention Page](https://grafana.com/docs/loki/v2.9.x/operations/storage/retention/).

## TLDR

- Modify `compactor.enabled` from `false` to `true` to deploy the compactor statefulset.
- Set `loki.config.compactor.retention_enabled` to `true`.
- Set `loki.config.compactor.limits_config.retention_period` to your desired value (e.g., `7d`).

## Solution

By default, compactor is disabled in loki-distributed chart. So first of all, enable the compactor statefulset in the loki-distributed chart.

```yaml
# charts/loki-distributed/values.yaml
compactor:
  enabled: true
```

I am using boltdb-shipper for index storage and a filesystem mounted with PV of gp3 type for chunk storage.

```console
$ kubectl get pvc -n grafana -o=jsonpath='{.items[*].spec.volumeName}' | xargs -n1 kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                       STORAGECLASS   REASON   AGE
pvc-4e05529b-7ae4-46dc-bc81-9f47d00dbe1c   10Gi       RWO            Delete           Bound    grafana/data-loki-distributed-compactor-0   gp3                     81m
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                      STORAGECLASS   REASON   AGE
pvc-797e97ab-62b7-4d94-86ac-1e50cb7b330d   30Gi       RWO            Delete           Bound    grafana/data-loki-distributed-ingester-0   gp3                     4h54m
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                      STORAGECLASS   REASON   AGE
pvc-62448f5f-9a8e-4903-8412-ffca13c720a7   30Gi       RWO            Delete           Bound    grafana/data-loki-distributed-ingester-1   gp3                     117m
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                      STORAGECLASS   REASON   AGE
pvc-2ef08f10-0a32-43a2-bbe2-e4a998838e23   30Gi       RWO            Delete           Bound    grafana/data-loki-distributed-ingester-2   gp3                     80m
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                     STORAGECLASS   REASON   AGE
pvc-f91c89e4-853d-4e76-882b-64ecea5bbf8e   10Gi       RWO            Delete           Bound    grafana/data-loki-distributed-querier-0   gp3                     3d3h
```

The relevant storage configuration can be found `Values.loki` in helm chart:

```yaml
# charts/loki-distributed/values.yaml
loki:
  # -- Check https://grafana.com/docs/loki/latest/configuration/#schema_config for more info on how to configure schemas
  schemaConfig:
    configs:
    - from: "2020-09-07"
      store: boltdb-shipper
      object_store: filesystem
      schema: v11
      index:
        prefix: loki_index_
        period: 24h

  # -- Check https://grafana.com/docs/loki/latest/configuration/#storage_config for more info on how to configure storages
  storageConfig:
    boltdb_shipper:
      shared_store: filesystem
      active_index_directory: /var/loki/index
      cache_location: /var/loki/cache
      cache_ttl: 168h
    filesystem:
      directory: /var/loki/chunks
```

> [!NOTE]
> According to [the official Loki documentation](https://grafana.com/docs/loki/v2.9.x/operations/storage/table-manager/#retention) in loki `v2.9.x`, Retention is only available if the index period is `24h`. Single store TSDB and single store BoltDB require 24h index period due to internal implementation.

Next, add the compactor settings to `loki.config` as shown below. By default, the compactor component is disabled in the loki-distributed chart.

For more details on configuring log retention, refer to this link.

Below is the compactor chart configuration with log retention applied.

```yaml
# charts/loki-distributed/values.yaml
loki:
  config: |
    # ... truncated
    compactor:
      shared_store: filesystem
      working_directory: /var/loki/compactor
      compaction_interval: 10m
      retention_enabled: true
      retention_delete_delay: 2h
      retention_delete_worker_count: 150
      delete_request_store: filesystem
    
    limits_config:
      retention_period: 7d
```

Finally, verify the injected Loki configuration file within the pod.

```bash
kubectl exec -it -n grafana loki-distributed-compactor-0 -- cat /etc/loki/config/config.yaml
```

`kubectl exec` output:

```yaml
# /etc/loki/config/config.yaml file from loki-distributed-compactor-0 pod
...
compactor:
  compaction_interval: 10m
  delete_request_store: filesystem
  retention_delete_delay: 2h
  retention_delete_worker_count: 150
  retention_enabled: true
  shared_store: filesystem
  working_directory: /var/loki/compactor
...
limits_config:
  retention_period: 7d
```

You can see that the log retention settings and `limits_config.retention_period` have been successfully added as intended.

## References

### Loki official docs

- [Retention](https://grafana.com/docs/loki/v2.9.x/operations/storage/retention/#table-manager) <sup>Loki v2.9.x</sup>

**Comment by arjunbhut** [comment]
@younsl I tried the above config with simplescalabe deployment, loki : v3.1.0, helm chart version : 6.10.0

`
loki:
  schemaConfig:
    configs:
      - from: 2024-04-01
        store: tsdb
        object_store: s3
        schema: v13
        index:
          prefix: loki_index_
          period: 24h
  ingester:
    chunk_encoding: snappy
  tracing:
    enabled: true
  querier:
    # Default is 4, if you have enough memory and CPU you can increase, reduce if OOMing
    max_concurrent: 4
  auth_enabled: false
  commonConfig:
    replication_factor: 1
  limits_config:
    retention_period: 48h
    max_line_size: 1048576
  compactor:
    working_directory: /var/loki/compactor
    compaction_interval: 5m
    retention_enabled: true
    retention_delete_delay: 2h
    retention_delete_worker_count: 150
    delete_request_store: s3

#gateway:
#  ingress:
#    enabled: true
#    hosts:
#      - host: FIXME
#        paths:
#          - path: /
#            pathType: Prefix

deploymentMode: SimpleScalable

backend:
  replicas: 3
read:
  replicas: 3
write:
  replicas: 3

# Enable minio for storage
minio:
  enabled: true

chunksCache:
  enabled: false

resultsCache:
  enable: false

memcachedExporter:
  enable: false

lokiCanary:
  enable: false

ruler:
  # -- The ruler component is optional and can be disabled if desired.
  enabled: true

# Zero out replica counts of other deployment modes
singleBinary:
  replicas: 0

ingester:
  replicas: 0
querier:
  replicas: 0
queryFrontend:
  replicas: 0
queryScheduler:
  replicas: 0
distributor:
  replicas: 0
# compactor:
#   enabled: true
compactor:
  # replicas: 0
  enabled: true
  # persistence:
  #   enabled: true
  #   storageClassName: worker-lvm-vg0  # Replace with your storage class name
   
  #   size: 10Gi 
indexGateway:
  replicas: 0
bloomCompactor:
  replicas: 0
bloomGateway:
  replicas: 0
`

But when i see the logs of loki-backend
i can see below log line , now i am not sure if retention is applied or not
`level=info ts=2024-09-22T19:27:16.328701733Z caller=expiration.go:78 msg="overall smallest retention period 1726860436.328, default smallest retention period 1726860436.328"`


Can you please help me with this

**Comment by younsl** [comment]
@arjunbhut Unfortunately, [Log retention](https://grafana.com/docs/loki/latest/operations/storage/retention/) page of Loki's official documentation does not provide a verification method, but only a setting method is mentioned.

What is crucial to note is that the log retention policy was effectively implemented and verified. After the configured retention period of 7 days had passed, older logs were indeed deleted as expected. This was observed in my `loki-distributed` system (Loki v2.9.8) using S3 as the object storage backend. The successful deletion of logs older than 7 days confirms that our retention settings are functioning correctly in this specific setup.

loki-distributed chart with log retention:

```yaml
# charts/loki-distributed/values.yaml (chartVersion: 0.79.2, loki appVersion: 2.9.8)
loki:
  config: |
    limits_config:
      retention_period: 7d

    compactor:
      shared_store: s3
      working_directory: /var/loki/compactor
      compaction_interval: 10m
      retention_enabled: true
      retention_delete_delay: 2h
      retention_delete_worker_count: 150
      delete_request_store: s3

  schemaConfig:
    configs:
    - from: "2024-08-01"
      store: tsdb
      object_store: s3
      schema: v13
      index:
        prefix: loki_index_
        # Note: Retention is only available if the index `period` is `24h`
        period: 24h

  storageConfig:
    tsdb_shipper:
      cache_ttl: 168h
    aws:
      s3: s3://ap-northeast-2
      bucketnames: <REDACTED>
```

During log retention process, it is a good idea to monitor compactor pod's logs and metrics to track the compactor's operating status and the amount of data deleted. It's also helpful to regularly check your object storage usage to see if it decreases over time.
I observed the total capacity of S3 storage and Compactor log after 7 days. After applying log retention for compactor, you need to wait for some time to verify if the retention policy is working properly. The reasons for this are as follows:

1. **Set retention period**: In the current configuration, the retention period is set to 7 days (`retention_period: 7d`). Therefore, it will take at least 7 days before old logs start to be deleted.
2. **Compactor operation cycle**: The compactor is set to operate every 5 minutes with `compaction_interval: 10m`. However, actual data deletion starts only after the retention period has passed.
3. **Deletion delay**: There's a 2-hour delay before actual deletion occurs, as set by `retention_delete_delay: 2h`, even after the retention period has passed.
4. **Existing data**: Depending on the amount of old log data already in the system, it may take longer for all data to be cleaned up.

I chose to wait until the log retention time passed. Other than this, I couldn't find a neat verification method.
